File: crates/ngb-db/src/groups.rs
use std::collections::HashMap;

use ngb_types::{NanoGridBotError, RegisteredGroup, Result};

use crate::connection::Database;

/// Repository for group storage and retrieval.
pub struct GroupRepository<'a> {
    db: &'a Database,
}

impl<'a> GroupRepository<'a> {
    pub fn new(db: &'a Database) -> Self {
        Self { db }
    }

    /// Save or update a group (upsert).
    pub async fn save_group(&self, group: &RegisteredGroup) -> Result<()> {
        let container_config = group
            .container_config
            .as_ref()
            .map(|c| serde_json::to_string(c).unwrap_or_default());

        sqlx::query(
            "INSERT OR REPLACE INTO groups
             (jid, name, folder, trigger_pattern, container_config, requires_trigger)
             VALUES (?, ?, ?, ?, ?, ?)",
        )
        .bind(&group.jid)
        .bind(&group.name)
        .bind(&group.folder)
        .bind(&group.trigger_pattern)
        .bind(&container_config)
        .bind(group.requires_trigger as i32)
        .execute(self.db.pool())
        .await
        .map_err(|e| NanoGridBotError::Database(format!("Save group: {e}")))?;

        Ok(())
    }

    /// Get a group by JID.
    pub async fn get_group(&self, jid: &str) -> Result<Option<RegisteredGroup>> {
        let row: Option<GroupRow> = sqlx::query_as(
            "SELECT jid, name, folder, trigger_pattern, container_config, requires_trigger
             FROM groups WHERE jid = ?",
        )
        .bind(jid)
        .fetch_optional(self.db.pool())
        .await
        .map_err(|e| NanoGridBotError::Database(format!("Get group: {e}")))?;

        Ok(row.map(row_to_group))
    }

    /// Get all registered groups.
    pub async fn get_all(&self) -> Result<Vec<RegisteredGroup>> {
        let rows: Vec<GroupRow> = sqlx::query_as(
            "SELECT jid, name, folder, trigger_pattern, container_config, requires_trigger
             FROM groups ORDER BY name ASC",
        )
        .fetch_all(self.db.pool())
        .await
        .map_err(|e| NanoGridBotError::Database(format!("Get all groups: {e}")))?;

        Ok(rows.into_iter().map(row_to_group).collect())
    }

    /// Get groups by folder.
    pub async fn get_by_folder(&self, folder: &str) -> Result<Vec<RegisteredGroup>> {
        let rows: Vec<GroupRow> = sqlx::query_as(
            "SELECT jid, name, folder, trigger_pattern, container_config, requires_trigger
             FROM groups WHERE folder = ? ORDER BY name ASC",
        )
        .bind(folder)
        .fetch_all(self.db.pool())
        .await
        .map_err(|e| NanoGridBotError::Database(format!("Get groups by folder: {e}")))?;

        Ok(rows.into_iter().map(row_to_group).collect())
    }

    /// Delete a group by JID.
    pub async fn delete_group(&self, jid: &str) -> Result<bool> {
        let result = sqlx::query("DELETE FROM groups WHERE jid = ?")
            .bind(jid)
            .execute(self.db.pool())
            .await
            .map_err(|e| NanoGridBotError::Database(format!("Delete group: {e}")))?;

        Ok(result.rows_affected() > 0)
    }

    /// Check if a group exists.
    pub async fn exists(&self, jid: &str) -> Result<bool> {
        let row: Option<(i32,)> = sqlx::query_as("SELECT 1 FROM groups WHERE jid = ?")
            .bind(jid)
            .fetch_optional(self.db.pool())
            .await
            .map_err(|e| NanoGridBotError::Database(format!("Check group exists: {e}")))?;

        Ok(row.is_some())
    }
}

#[derive(sqlx::FromRow)]
struct GroupRow {
    jid: String,
    name: String,
    folder: String,
    trigger_pattern: Option<String>,
    container_config: Option<String>,
    requires_trigger: i32,
}

fn row_to_group(row: GroupRow) -> RegisteredGroup {
    let container_config: Option<HashMap<String, serde_json::Value>> = row
        .container_config
        .as_deref()
        .and_then(|s| serde_json::from_str(s).ok());

    RegisteredGroup {
        jid: row.jid,
        name: row.name,
        folder: row.folder,
        trigger_pattern: row.trigger_pattern,
        container_config,
        requires_trigger: row.requires_trigger != 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::connection::Database;

    async fn setup() -> Database {
        let db = Database::in_memory().await.unwrap();
        db.initialize().await.unwrap();
        db
    }

    fn make_group(jid: &str, name: &str, folder: &str) -> RegisteredGroup {
        RegisteredGroup {
            jid: jid.to_string(),
            name: name.to_string(),
            folder: folder.to_string(),
            trigger_pattern: None,
            container_config: None,
            requires_trigger: true,
        }
    }

    #[tokio::test]
    async fn save_and_get_group() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        let group = make_group("tg:123", "Test Group", "test");
        repo.save_group(&group).await.unwrap();

        let found = repo.get_group("tg:123").await.unwrap().unwrap();
        assert_eq!(found.jid, "tg:123");
        assert_eq!(found.name, "Test Group");
        assert!(found.requires_trigger);
    }

    #[tokio::test]
    async fn save_group_upsert() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        let mut group = make_group("tg:123", "Old Name", "test");
        repo.save_group(&group).await.unwrap();

        group.name = "New Name".to_string();
        repo.save_group(&group).await.unwrap();

        let found = repo.get_group("tg:123").await.unwrap().unwrap();
        assert_eq!(found.name, "New Name");
    }

    #[tokio::test]
    async fn get_all_groups() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        repo.save_group(&make_group("a:1", "Alpha", "alpha"))
            .await
            .unwrap();
        repo.save_group(&make_group("b:2", "Beta", "beta"))
            .await
            .unwrap();

        let all = repo.get_all().await.unwrap();
        assert_eq!(all.len(), 2);
        // Sorted by name
        assert_eq!(all[0].name, "Alpha");
        assert_eq!(all[1].name, "Beta");
    }

    #[tokio::test]
    async fn get_by_folder() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        repo.save_group(&make_group("a:1", "A", "shared"))
            .await
            .unwrap();
        repo.save_group(&make_group("b:2", "B", "shared"))
            .await
            .unwrap();
        repo.save_group(&make_group("c:3", "C", "other"))
            .await
            .unwrap();

        let shared = repo.get_by_folder("shared").await.unwrap();
        assert_eq!(shared.len(), 2);
    }

    #[tokio::test]
    async fn delete_group() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        repo.save_group(&make_group("tg:1", "G1", "f1"))
            .await
            .unwrap();
        assert!(repo.exists("tg:1").await.unwrap());

        let deleted = repo.delete_group("tg:1").await.unwrap();
        assert!(deleted);
        assert!(!repo.exists("tg:1").await.unwrap());

        let deleted_again = repo.delete_group("tg:1").await.unwrap();
        assert!(!deleted_again);
    }

    #[tokio::test]
    async fn group_not_found() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        let found = repo.get_group("nonexistent").await.unwrap();
        assert!(found.is_none());
    }

    #[tokio::test]
    async fn group_with_container_config() {
        let db = setup().await;
        let repo = GroupRepository::new(&db);

        let mut config = HashMap::new();
        config.insert(
            "timeout".to_string(),
            serde_json::Value::Number(serde_json::Number::from(60)),
        );

        let mut group = make_group("tg:1", "G1", "f1");
        group.container_config = Some(config);
        repo.save_group(&group).await.unwrap();

        let found = repo.get_group("tg:1").await.unwrap().unwrap();
        let cc = found.container_config.unwrap();
        assert_eq!(cc.get("timeout").unwrap(), &serde_json::json!(60));
    }
}

File: src/nanogridbot/database/groups.py
"""Group database operations."""

import json
from collections.abc import Sequence
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from nanogridbot.database.connection import Database

from nanogridbot.types import RegisteredGroup


class GroupRepository:
    """Repository for group storage and retrieval."""

    def __init__(self, database: "Database") -> None:
        """Initialize group repository.

        Args:
            database: Database connection instance.
        """
        self._db = database

    async def save_group(self, group: RegisteredGroup) -> None:
        """Save or update a group.

        Args:
            group: Group to save.
        """
        container_config = (
            json.dumps(group.container_config) if group.container_config is not None else None
        )

        await self._db.execute(
            """
            INSERT OR REPLACE INTO groups
            (jid, name, folder, user_id, trigger_pattern, container_config, requires_trigger)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (
                group.jid,
                group.name,
                group.folder,
                group.user_id,
                group.trigger_pattern,
                container_config,
                int(group.requires_trigger),
            ),
        )
        await self._db.commit()

    async def get_group(self, jid: str) -> RegisteredGroup | None:
        """Get a group by JID.

        Args:
            jid: Group JID.

        Returns:
            Group if found, None otherwise.
        """
        row = await self._db.fetchone(
            """
            SELECT jid, name, folder, user_id, trigger_pattern, container_config, requires_trigger
            FROM groups
            WHERE jid = ?
            """,
            (jid,),
        )
        return self._row_to_group(row) if row else None

    async def get_groups(self) -> Sequence[RegisteredGroup]:
        """Get all registered groups.

        Returns:
            List of registered groups.
        """
        rows = await self._db.fetchall(
            """
            SELECT jid, name, folder, user_id, trigger_pattern, container_config, requires_trigger
            FROM groups
            ORDER BY name ASC
            """,
        )
        return [self._row_to_group(row) for row in rows]

    async def get_groups_by_folder(self, folder: str) -> Sequence[RegisteredGroup]:
        """Get groups by folder name.

        Args:
            folder: Folder name to filter by.

        Returns:
            List of groups in the folder.
        """
        rows = await self._db.fetchall(
            """
            SELECT jid, name, folder, user_id, trigger_pattern, container_config, requires_trigger
            FROM groups
            WHERE folder = ?
            ORDER BY name ASC
            """,
            (folder,),
        )
        return [self._row_to_group(row) for row in rows]

    async def get_groups_by_user(self, user_id: int) -> Sequence[RegisteredGroup]:
        """Get groups by user ID.

        Args:
            user_id: User ID to filter by.

        Returns:
            List of groups owned by the user.
        """
        rows = await self._db.fetchall(
            """
            SELECT jid, name, folder, user_id, trigger_pattern, container_config, requires_trigger
            FROM groups
            WHERE user_id = ?
            ORDER BY name ASC
            """,
            (user_id,),
        )
        return [self._row_to_group(row) for row in rows]

    async def delete_group(self, jid: str) -> bool:
        """Delete a group by JID.

        Args:
            jid: Group JID to delete.

        Returns:
            True if deleted, False if not found.
        """
        cursor = await self._db.execute(
            "DELETE FROM groups WHERE jid = ?",
            (jid,),
        )
        await self._db.commit()
        return cursor.rowcount > 0

    async def group_exists(self, jid: str) -> bool:
        """Check if a group exists.

        Args:
            jid: Group JID to check.

        Returns:
            True if exists, False otherwise.
        """
        row = await self._db.fetchone(
            "SELECT 1 FROM groups WHERE jid = ?",
            (jid,),
        )
        return row is not None

    @staticmethod
    def _row_to_group(row: dict[str, Any]) -> RegisteredGroup:
        """Convert database row to RegisteredGroup model.

        Args:
            row: Database row dictionary.

        Returns:
            RegisteredGroup instance.
        """
        container_config_raw = row.get("container_config")
        container_config: dict[str, Any] | None = None
        if container_config_raw:
            try:
                container_config = json.loads(container_config_raw)
            except json.JSONDecodeError:
                pass

        return RegisteredGroup(
            jid=str(row["jid"]),
            name=str(row["name"]),
            folder=str(row["folder"]),
            user_id=row["user_id"] if row.get("user_id") else None,
            trigger_pattern=row["trigger_pattern"] if row["trigger_pattern"] else None,
            container_config=container_config,
            requires_trigger=bool(row["requires_trigger"]),
        )
